src/optparser/optSpec.ml

```diff
@@ -1,71 +1,79 @@
-(**
-   A module for the functions that operate on an option and  a set of options.
-   For example,
-   - functions that lift option info into an option
-   - functions that merge option sets
-   - etc.
-   are included.
-   @author Clare Jang
- *)
-
-(**
-   An error type that might be emitted from {!Parser.parse} function.
-   @author Clare Jang
- *)
-type error =
-  | MissingMandatory
-    of string (* option name *)
-  | InvalidArgLength
-    of string (* option name *)
-       * int (* expected number of arguments *)
-       * int (* actual number of arguments *)
-  | ArgReaderFailure
-    of string (* option name *)
-  | NotAnOption
-    of string (* option name *)
-
-type help_entry =
-  OptName.t (* option name *)
-  * string list (* names for arguments *)
-  * string option (* help message for option *)
+(** A module for the functions that operate on an option and a set of
+    options. For example,
+
+    - functions that lift option info into an option
+    - functions that merge option sets
+    - etc. are included.
+
+    @author Clare Jang *)
+
+module Error = struct
+  module Option = struct
+    type option_error = { option_name : string }
+  end
+
+  module Argument = struct
+    type invalid_arguments_length_error =
+      { option_name : string
+      ; expected_argument_count : int
+      ; actual_argument_count : int
+      }
+  end
+
+  type t =
+    [ `Missing_mandatory_option of Option.option_error
+    | `Invalid_arguments_length of Argument.invalid_arguments_length_error
+    | `Argument_reader_failure of Option.option_error
+    | `Not_an_option of Option.option_error
+    ]
```

Comment:
May I ask the reason for this change? (Normal variant to polymorphic variant)

Reply:
That error type is being used with `('a, 'e) result`. Using polymorphic variants allows for composable error-handling: https://keleshev.com/composable-error-handling-in-ocaml

I realize that composable errors are not actually used in `Optparser`. To support that, we would need to make these changes to the parser types:
```diff
(* parser.mli *)
- val parse : 'a OptSpec.t -> string list -> ('a, OptSpec.Error.t) result
+ val parse : 'a OptSpec.t -> string list -> ('a, [> OptSpec.Error.t]) result

(* optSpec.ml *)
- type +'a t =
+ type (+'a, 'error) t 
  { opt_tbl :
      (string, int option * (help_printer -> string list -> unit)) Hashtbl.t
-  ; comp_value : string list -> ('a, Error.t) result
+  ; comp_value : string list -> ('a, [> Error.t] as 'error) result
  ; mandatory_help_entries : HelpEntry.t list
  ; optional_help_entries : HelpEntry.t list
  }
```